pkg/query-service/app/metrics/v4/helpers/sub_query.go

```diff
@@ -13,31 +13,129 @@ import (
+func WhichSamplesTableToUse(start, end int64, mq *v3.BuilderQuery) string {
+
+	// we don't have any aggregated table for sketches (yet)
+	if mq.AggregateAttribute.Type == v3.AttributeKeyType(v3.MetricTypeExponentialHistogram) {
+		return constants.SIGNOZ_EXP_HISTOGRAM_TABLENAME
+	}
+
+	// if the time aggregation is count_distinct, we need to use the distributed_samples_v4 table
+	// because the aggregated tables don't support count_distinct
+	if mq.TimeAggregation == v3.TimeAggregationCountDistinct {
+		return constants.SIGNOZ_SAMPLES_V4_TABLENAME
+	}
+
+	if end-start < oneDayInMilliseconds {
+		return constants.SIGNOZ_SAMPLES_V4_TABLENAME
+	} else if end-start < oneWeekInMilliseconds {
+		return constants.SIGNOZ_SAMPLES_V4_AGG_5M_TABLENAME
+	} else {
+		return constants.SIGNOZ_SAMPLES_V4_AGG_30M_TABLENAME
+	}
+}
+
+func AggregationColumnForSamplesTable(start, end int64, mq *v3.BuilderQuery) string {
+	tableName := WhichSamplesTableToUse(start, end, mq)
+	var aggregationColumn string
+	switch mq.Temporality {
+	case v3.Delta:
+		// for delta metrics, we only support `RATE`/`INCREASE` both of which are sum
+		switch tableName {
+		case constants.SIGNOZ_SAMPLES_V4_TABLENAME:
+			aggregationColumn = "sum(value)"
+		case constants.SIGNOZ_SAMPLES_V4_AGG_5M_TABLENAME, constants.SIGNOZ_SAMPLES_V4_AGG_30M_TABLENAME:
+			aggregationColumn = "sum(sum)"
+		}
+	case v3.Cumulative:
+		// for cumulative metrics, we only support `RATE`/`INCREASE`. The max value in window is
+		// used to calculate the sum which is then divided by the window size to get the rate
+		switch tableName {
+		case constants.SIGNOZ_SAMPLES_V4_TABLENAME:
+			aggregationColumn = "max(value)"
+		case constants.SIGNOZ_SAMPLES_V4_AGG_5M_TABLENAME, constants.SIGNOZ_SAMPLES_V4_AGG_30M_TABLENAME:
+			aggregationColumn = "max(max)"
+		}
+	case v3.Unspecified:
+		switch tableName {
+		case constants.SIGNOZ_SAMPLES_V4_TABLENAME:
+			switch mq.TimeAggregation {
+			case v3.TimeAggregationAnyLast:
+				aggregationColumn = "anyLast(value)"
+			case v3.TimeAggregationSum:
+				aggregationColumn = "sum(value)"
+			case v3.TimeAggregationAvg:
+				aggregationColumn = "avg(value)"
+			case v3.TimeAggregationMin:
+				aggregationColumn = "min(value)"
+			case v3.TimeAggregationMax:
+				aggregationColumn = "max(value)"
+			case v3.TimeAggregationCount:
+				aggregationColumn = "count(value)"
+			case v3.TimeAggregationCountDistinct:
+				aggregationColumn = "countDistinct(value)"
+			case v3.TimeAggregationRate, v3.TimeAggregationIncrease: // ideally, this should never happen
+				aggregationColumn = "sum(value)"
+			}
+		case constants.SIGNOZ_SAMPLES_V4_AGG_5M_TABLENAME, constants.SIGNOZ_SAMPLES_V4_AGG_30M_TABLENAME:
+			switch mq.TimeAggregation {
+			case v3.TimeAggregationAnyLast:
+				aggregationColumn = "anyLast(last)"
+			case v3.TimeAggregationSum:
+				aggregationColumn = "sum(sum)"
+			case v3.TimeAggregationAvg:
+				aggregationColumn = "sum(sum) / sum(count)"
```

Comment:
more efficient to store than to do twice the work in query?

Reply:
This would be row level operation, so should be good I guess

Reply:
The other way would be storing the `avgState` which internally does the [same](https://github.com/ClickHouse/ClickHouse/blob/ae5223854f71361a4cafdd6bafad3e2cc770432b/src/AggregateFunctions/AggregateFunctionAvg.h#L104-L128) and `-Merge` at the query time. There won't be any noticeable difference.